File: cil-tools/ext3.c
/* This is an example import file, showing how to import the ext3
 * superblock automatically from Linux header files.
 *
 * Use: bitstring-import-c ext3.c > ext3.bmpp
 *
 * Tip: Add the --debug flag to that command line to see what's going on.
 * Also use bitstring-objinfo to examine the bmpp file.
 */

/* These are required by Linux in order to get the little/big-endian
 * notations present in the Linux kernel header files.  Any other
 * defines needed can go here.
 */
#define __CHECKER__      1
#define __CHECK_ENDIAN__ 1

/* Include files necessary to get the structure(s) and constant(s) we're
 * interested in.
 */
#include <linux/magic.h>
#include <linux/ext2_fs.h>

/* This tells the importer program what structures and constants to import. */
//typedef struct ext3_super_block BITSTRING_IMPORT(ext3_super_block);
typedef struct ext2_super_block BITSTRING_IMPORT(ext3_super_block);
BITSTRING_CONSTANT_INT32 (ext3_super_magic, EXT3_SUPER_MAGIC);

File: cil-tools/ext3.ml
(* This is an example program using an imported C structure.
 *
 * The C structure is imported from <linux/ext3_fs.h> via "ext3.c"
 * by the bitstring-import-c program, and saved as "ext3.bmpp".
 *
 * Then we can load "ext3.bmpp" here.
 *)

open Printf

open bitmatch "ext3.bmpp"

let () =
  (* Load a real ext3 superblock from the examples directory. *)
  let bits = Bitstring.bitstring_of_file "examples/ext3_sb" in

  bitmatch bits with
  | { :ext2_super_block } ->
      printf "ext3 superblock:\n";
      printf "  s_inodes_count = %ld\n" s_inodes_count;
      printf "  s_blocks_count = %ld\n" s_blocks_count;
      printf "  s_free_inodes_count = %ld\n" s_free_inodes_count;
      printf "  s_free_blocks_count = %ld\n" s_free_blocks_count;
      printf "  s_uuid = %S\n" s_uuid;
      printf "  s_volume_name = %S\n" s_volume_name;
      printf "  s_last_mounted = %S\n" s_last_mounted

  | { _ } ->
      failwith "input is not an ext2/3 superblock"
